pkg/gadgets/snisnoop/tracer/tracer.go

```diff
@@ -0,0 +1,196 @@
+	"sync"
+	"syscall"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/perf"
+	"golang.org/x/sys/unix"
+
+	"github.com/kinvolk/inspektor-gadget/pkg/rawsock"
+)
+
+// #include "bpf/snisnoop.h"
+import "C"
+
+const (
+	BPF_PROG_NAME = "bpf_prog1"
+	BPF_MAP_NAME  = "events"
+	SO_ATTACH_BPF = 50
+)
+
+type link struct {
+	collection *ebpf.Collection
+	perfRd     *perf.Reader
+
+	sockFd int
+
+	// users count how many users called Attach(). This can happen for two reasons:
+	// 1. several containers in a pod (sharing the netns)
+	// 2. pods with networkHost=true
+	users int
+}
+
+type Tracer struct {
+	mu sync.Mutex
+
+	spec *ebpf.CollectionSpec
+
+	// key: namespace/podname
+	// value: Tracelet
+	attachments map[string]*link
+}
+
+func NewTracer() (*Tracer, error) {
+	spec, err := ebpf.LoadCollectionSpecFromReader(bytes.NewReader(ebpfProg))
+	if err != nil {
+		return nil, fmt.Errorf("failed to load asset: %w", err)
+	}
+
+	t := &Tracer{
+		spec:        spec,
+		attachments: make(map[string]*link),
+	}
+
+	return t, nil
+}
+
+func (t *Tracer) Attach(
+	key string,
+	pid uint32,
+	callback func(name string),
+	errFunc func(err error),
+) error {
+	if l, ok := t.attachments[key]; ok {
+		l.users++
+		return nil
+	}
+
+	coll, err := ebpf.NewCollectionWithOptions(
+		t.spec,
+		ebpf.CollectionOptions{
+			Programs: ebpf.ProgramOptions{
+				LogSize: ebpf.DefaultVerifierLogSize * 100,
+			},
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create BPF collection: %w", err)
+	}
+
+	rd, err := perf.NewReader(coll.Maps[BPF_MAP_NAME], os.Getpagesize())
+	if err != nil {
+		return fmt.Errorf("failed to get a perf reader: %w", err)
+	}
+
+	prog, ok := coll.Programs[BPF_PROG_NAME]
+	if !ok {
+		return fmt.Errorf("failed to find BPF program %q", BPF_PROG_NAME)
+	}
+
+	sockFd, err := rawsock.OpenRawSock(pid)
+	if err != nil {
+		return fmt.Errorf("failed to open raw socket: %w", err)
+	}
+
+	if err := syscall.SetsockoptInt(sockFd, syscall.SOL_SOCKET, SO_ATTACH_BPF, prog.FD()); err != nil {
+		return fmt.Errorf("failed to attach BPF program: %w", err)
+	}
+
+	l := &link{
+		collection: coll,
+		sockFd:     sockFd,
+		perfRd:     rd,
+		users:      1,
+	}
+	t.attachments[key] = l
+
+	go t.listen(key, rd, callback, errFunc)
+
+	return nil
+}
+
+func parseSNIEvent(rawSample []byte) (ret string) {
+	name := make([]byte, C.TLS_MAX_SERVER_NAME_LEN)
+	copy(name, rawSample)
+
+	str := string(name)
+	i := strings.Index(str, "\x00")
+	if i > 0 {
+		return str[:i]
+	}
+	return str
+}
+
+func (t *Tracer) listen(
+	key string,
+	rd *perf.Reader,
+	callback func(name string),
+	errFunc func(err error),
+) {
+	for {
+		record, err := rd.Read()
+		if err != nil {
+			if perf.IsClosed(err) {
+				return
+			}
+			errFunc(fmt.Errorf("Error while reading from perf event reader (%s): %s", key, err))
+			return
+		}
+
+		if record.LostSamples != 0 {
+			errFunc(fmt.Errorf("Warning: perf event ring buffer full, dropped %d samples (%s)", record.LostSamples, key))
```

Comment:
Nit: The type will be sent as `ERR`, but it's actually a warning. (I think it doesn't matter that much now because the client will just print a message anyway).

Reply:
Instead of adding a 3rd callback for warning messages, I'll try to use `pkg/types` as you did in execsnoop.